### ISSUE TITLE
Don't show legend for size with scale.zero = true?

### ISSUE THREAD

**Issue by kanitw** [vega/vega#717]
Vega-Lite outputs always have `zero: true` for scale of `size` to ensure correct scaling.  

However, when `zero` is `true`, Vega legend always include 0, which looks kinda weird. 

For example, in this [VG spec](http://pastebin.com/uswTKN4P) from the following VL spec:
```
{
  "description": "A simple bar chart with embedded data.",
  "data": {
    "values": [
      {"a": "A","b": 10},
      {"a": "B","b": 20},
      {"a": "C","b": 30},
      {"a": "D","b": 40}
    ]
  },
  "mark": "circle",
  "encoding": {
    "x": {"field": "a", "type": "ordinal"},
    "size": {"field": "b", "type": "quantitative", "scale": {"zero": true, "range": [0, 361]}}
  }
}
```

![vega_editor](https://cloud.githubusercontent.com/assets/111269/21968365/ca58d9ec-db48-11e6-91b3-5af739ecd493.png)

The field b spans [10,40], but the legend include 0, which is unnecessary.  

It'd be great if 0 is excluded.

**Comment by jheer** [comment]
I don't think excluding zero is always a good idea. However, I could imagine adding an extra property (`excludeZero`?) to legend definitions that lets you explicitly remove zero from the generated legend entries. This feels a bit hacky to me, but I don't have a preferred solution at the moment. Would that work for you?

**Comment by jheer** [comment]
Also: as you think about this, what do you want to do when the size-encoded data includes negative values? In that case it seems reasonable to keep the zero entry.

**Comment by kanitw** [comment]
> I don't think excluding zero is always a good idea. 

> what do you want to do when the size-encoded data includes negative values? In that case it seems reasonable to keep the zero entry.

I agree that "excluding zero is always a good idea". 

I think a reasonable rule is zero size should be included in the legend if the specified `domain` prior to applying `zero` flag has already included zero (i.e., `domainMin` <= 0).   

Note that, in such cases, the `zero` flag does not affect the domain.

> However, I could imagine adding an extra property (excludeZero?) to legend definitions that lets you explicitly remove zero from the generated legend entries. 

This works for me, but I wonder if legend has knowledge about the scale domain (prior to applying `zero` flag).  If so, maybe we don't need this hacky flag?  

(I can't think of case where `domainMin` > 0 and we need to show zero in the legend.) 

On a related note, *later* we might want to add a way to throw warning if users map negative values to size as it's a bad encoding (https://github.com/vega/vega-lite/issues/1833).  That said, it seems complicated and probably require data quality assessment operator in Vega, so let's not worry about this for now.

**Comment by domoritz** [comment]
> I agree that "excluding zero is always a good idea".

You mean "I agree that "excluding zero is **not** always a good idea".", right?

**Comment by kanitw** [comment]
Ah yes, I'll correct it above!

**Comment by g3o2** [comment]
> The field b spans [10,40], but the legend include 0, which is unnecessary.
>
> It'd be great if 0 is excluded.

According to the Vega 2 [scale documentation](https://github.com/vega/vega/wiki/Scales), setting `"zero": true`" ensures that a zero baseline value is included in the scale domain", so that the behaviour to include 0 in the legend is consistent, albeit weird looking.

Possible workarounds:

* setting `rangeMin` to 1, so that `zero` will be visible by a dot but this also introduces a minor distortion;
* setting `"zero": false` together with a `rangeMin` higher than 0;
* setting the `legend.values` manually (not practical). 

For now, `legend.values` property only accepts an array, making it accept a `DataRef` may provide a more convenient and general purpose solution to the above issue.

**Comment by jheer** [comment]
Right now I'm experimenting with the following and so far it seems to work well: Remove a zero-sized entry if and only if (1) there is an active size encoding driving the legend, (2) the removed value is the first entry in the legend, and (3) the removed value was generated from the scale domain, not supplied as part of an explicit `values` array. Sound good?

(Note that our legend entry generator has access to the backing scales, but not their Vega specs, so we don't know if `zero: true` is set or not.)

**Comment by g3o2** [comment]
I would tend towards the more explicit `excludeZero` property (or any such property), for it reflects the user's conscious choice to do something that's not necessarily consistent (removing 0 from the legend) with a previous conscious decision (adding 0 to the scale domain). In a grammar of graphics, what makes sense or not should be the user's decision.

**Comment by jheer** [comment]
Suppression of initial zero-sized legend entries was included in the `v3.0.0-beta.15` release.
